pkg/topology/kademlia/kademlia.go

```diff
@@ -868,8 +874,15 @@ func (k *Kad) Announce(ctx context.Context, peer swarm.Address, fullnode bool) e
 				// about lightnodes to others.
 				continue
 			}
+			k.bgBroadcastWg.Add(1)
 			go func(connectedPeer swarm.Address) {
-				if err := k.discovery.BroadcastPeers(ctx, connectedPeer, peer); err != nil {
+				defer k.bgBroadcastWg.Done()
+
+				// Create a new deadline ctx to prevent goroutine pile up
+				cCtx, cCancel := context.WithTimeout(k.bgBroadcastCtx, time.Minute)
+				defer cCancel()
+
+				if err := k.discovery.BroadcastPeers(cCtx, connectedPeer, peer); err != nil {
```

Comment:
since the context which is passed to `Announce` is basically ignored, consider removing the context parameter from the signature and move the whole wg shebang directly into `hive`. Also, waiting on the goroutines on shutdown might not be critical in case they do not need to access any resource that may terminate on shutdown (statestore, etc), resulting in a panic and an inconsistent shutdown

Reply:
There is a 2-way broadcast here. The background broadcasts are for nodes that we know, and then there is a broadcast back to the connected peer that happens in the same context passed to Announce.

We won't be able to use the context from hive for broadcast peers in this case unless we remove the context parameter from the BroadcastPeers call as well. Or you are suggesting to use background context in the broadcast peers call and then using a timeout context in hive along with wg logic? So not have a background context to cancel?

BroadcastPeers uses p2p.NewStream, that will try to get a new connection. This is why I felt it would be better to shutdown cleanly.

Reply:
sorry i missed that. so it would be maybe more correct to use the parameter context for broadcasting to the single peer, but use a background context for broadcasting to the other peers

Reply:
Instead of using background context, I am using a global cancellable context, which is cancelled on stop, so it should cancel any in process broadcasts correctly. Better than using background context I feel.